moshi/src/main/java/com/squareup/moshi/ClassJsonAdapter.kt

```diff
@@ -0,0 +1,229 @@
+
+  override fun fromJson(reader: JsonReader): T {
+    val result: T = try {
+      classFactory.newInstance()
+    } catch (e: InstantiationException) {
+      throw RuntimeException(e)
+    } catch (e: InvocationTargetException) {
+      throw e.rethrowCause()
+    } catch (e: IllegalAccessException) {
+      throw AssertionError()
+    }
+
+    try {
+      reader.beginObject()
+      while (reader.hasNext()) {
+        val index = reader.selectName(options)
+        if (index == -1) {
+          reader.skipName()
+          reader.skipValue()
+          continue
+        }
+        fieldsArray[index].read(reader, result as Any)
+      }
+      reader.endObject()
+      return result
+    } catch (e: IllegalAccessException) {
+      throw AssertionError()
+    }
+  }
+
+  override fun toJson(writer: JsonWriter, value: T?) {
+    try {
+      writer.beginObject()
+      fieldsArray.forEach { fieldBinding ->
+        writer.name(fieldBinding.name)
+        fieldBinding.write(writer, value as Any)
+      }
+      writer.endObject()
+    } catch (e: IllegalAccessException) {
+      throw AssertionError()
+    }
+  }
+
+  override fun toString() = "JsonAdapter($classFactory)"
+
+  internal open class FieldBinding<T>(val name: String, val field: Field, val adapter: JsonAdapter<T>) {
+    open fun read(reader: JsonReader, value: Any) {
+      val fieldValue = adapter.fromJson(reader)
+      field[value] = fieldValue
+    }
+
+    @Suppress("UNCHECKED_CAST") // We require that field's values are of type T.
+    open fun write(writer: JsonWriter, value: Any) {
+      val fieldValue = field[value] as T
+      adapter.toJson(writer, fieldValue)
+    }
+  }
+
+  companion object Factory : JsonAdapter.Factory {
+    override fun create(
+      type: Type,
+      annotations: Set<Annotation>,
+      moshi: Moshi
+    ): JsonAdapter<*>? {
+      if (type !is Class<*> && type !is ParameterizedType) {
+        return null
+      }
+      val rawType = type.rawType
+      if (rawType.isInterface || rawType.isEnum) return null
+      if (annotations.isNotEmpty()) return null
+      if (rawType.isPlatformType) {
+        type.throwIfIsCollectionClass(List::class.java)
+        type.throwIfIsCollectionClass(Set::class.java)
+        type.throwIfIsCollectionClass(Map::class.java)
+        type.throwIfIsCollectionClass(Collection::class.java)
+
+        var messagePrefix = "Platform $rawType"
+        if (type is ParameterizedType) {
+          messagePrefix += " in $type"
+        }
+        throw IllegalArgumentException(
+          "$messagePrefix requires explicit JsonAdapter to be registered"
+        )
+      }
+
+      require(!rawType.isAnonymousClass) {
+        "Cannot serialize anonymous class ${rawType.name}"
+      }
+      require(!rawType.isLocalClass) {
+        "Cannot serialize local class ${rawType.name}"
+      }
+      require(!(rawType.enclosingClass != null && !Modifier.isStatic(rawType.modifiers))) {
```

Comment:
Please simplify this logic to avoid double-negatives

Reply:
https://github.com/square/moshi/pull/1470/commits/44c4488c549786794e573e19c0d4d6a3bc523593